spec.html

```diff
@@ -4278,73 +4278,91 @@ <h1>UpdateEmpty ( _completionRecord_, _value_ )</h1>
+          <tbody>
+            <tr>
+              <th>Field Name</th>
+              <th>Value</th>
+              <th>Meaning</th>
+            </tr>
+            <tr>
+              <td oldids="sec-getbase,ao-getbase">[[Base]]</td>
+              <td>
+                One of:
+                <ul>
+                  <li>
+                    ECMAScript language value except *undefined* and *null*;
+                  </li>
+                  <li>
+                    Environment Record;
+                  </li>
+                  <li>
+                    ~empty~.
+                  </li>
+                </ul>
+              </td>
+              <td>~empty~ value indicates that the Reference could not be resolved to a binding.</td>
+            </tr>
+            <tr>
+              <td oldids="sec-getreferencedname,ao-getreferencedname">[[ReferencedName]]</td>
+              <td>String or Symbol</td>
+              <td>Binding name. Always a String if [[Base]] value is an Environment Record.</td>
+            </tr>
+            <tr>
+              <td oldids="sec-isstrictreference,ao-isstrictreference">[[Strict]]</td>
+              <td>Boolean</td>
+              <td>*true* if the Reference originated in strict mode code, *false* otherwise.</td>
+            </tr>
+            <tr>
+              <td>[[ThisValue]]</td>
+              <td>any ECMAScript language value or ~empty~</td>
+              <td>If not ~empty~, the Reference represents a name binding that was expressed using the `super` keyword; it is called a <dfn id="super-reference">Super Reference</dfn> and its [[Base]] value will never be an Environment Record.</td>
+            </tr>
+          </tbody>
+        </table>
+      </emu-table>
 
-      <emu-clause id="sec-isstrictreference" aoid="IsStrictReference" oldids="ao-isstrictreference">
-        <h1>IsStrictReference ( _V_ )</h1>
-        <p>The abstract operation IsStrictReference takes argument _V_. It performs the following steps when called:</p>
-        <emu-alg>
-          1. Assert: Type(_V_) is Reference.
-          1. Return the strict reference flag of _V_.
-        </emu-alg>
-      </emu-clause>
+      <p>The following abstract operations are used in this specification to operate upon References:</p>
 
       <emu-clause id="sec-hasprimitivebase" aoid="HasPrimitiveBase" oldids="ao-hasprimitivebase">
         <h1>HasPrimitiveBase ( _V_ )</h1>
         <p>The abstract operation HasPrimitiveBase takes argument _V_. It performs the following steps when called:</p>
         <emu-alg>
-          1. Assert: Type(_V_) is Reference.
-          1. If Type(_V_'s base value component) is Boolean, String, Symbol, BigInt, or Number, return *true*; otherwise return *false*.
+          1. Assert: _V_ is a Reference Record.
+          1. If _V_.[[Base]] is ~empty~, then return *false*.
+          1. If Type(_V_.[[Base]]) is either Boolean, String, Symbol, BigInt, or Number, return *true*; otherwise return *false*.
```

Comment:
hm, why aren't null/undefined here?

Reply:
There is (as of this PR) no way to produce a Reference whose `[[Base]]` is either `null` or `undefined`, so it wouldn't make sense to include them.

(Prior to this PR, a base value component of `undefined` was used to indicate an unresolvable reference, that is, a reference to an undeclared global name. But there still wasn't a way to have a _property_ reference whose base was `undefined` or `null`, because all of the paths which might have produced such a reference are guarded by `RequireObjectCoercible`.)